File: app/src/main/java/workshop/akbolatss/tagsnews/screen/recommendations/RecommendationsFragment.java
package workshop.akbolatss.tagsnews.screen.recommendations;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import java.util.List;
import java.util.Random;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnFocusChange;
import butterknife.OnTextChanged;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.subjects.BehaviorSubject;
import workshop.akbolatss.tagsnews.R;
import workshop.akbolatss.tagsnews.application.App;
import workshop.akbolatss.tagsnews.di.component.AppComponent;
import workshop.akbolatss.tagsnews.di.component.DaggerSourcesComponent;
import workshop.akbolatss.tagsnews.di.module.SourcesModule;
import workshop.akbolatss.tagsnews.repositories.source.RssSource;
import workshop.akbolatss.tagsnews.screen.sources.SourcesPresenter;
import workshop.akbolatss.tagsnews.screen.sources.SourcesView;

public class RecommendationsFragment extends Fragment implements SourcesView, RecommendationsAdapter.onSourceClickListener {

    private static final String TAG = "RecommendationsFragment";

    @Inject
    protected Context mContext;

    @Inject
    protected SourcesPresenter mPresenter;

    @BindView(R.id.tvNoContent)
    protected TextView tvNoContent;

    @BindView(R.id.recyclerView)
    protected RecyclerView mRecyclerV;
    private RecommendationsAdapter mAdapter;

    @BindView(R.id.btnSearch)
    protected ImageButton mBtnSearch;

    @BindView(R.id.progressBar)
    protected ProgressBar mProgress;

    @BindView(R.id.edit_query)
    protected EditText mEditText;

    private OnFragmentInteractionListener mListener;

    private BehaviorSubject<String> searchSubject;

    public RecommendationsFragment() {
        // Required empty public constructor
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_recommendations, container, false);
        ButterKnife.bind(this, rootView);
        initDagger();
        return rootView;
    }

    protected void initDagger() {
        DaggerSourcesComponent.builder()
                .appComponent(getAppComponent())
                .sourcesModule(new SourcesModule(this))
                .build()
                .inject(this);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        
        initRV();
        initEditText();
    }

    private void initRV() {
        mRecyclerV.setHasFixedSize(true);
        mRecyclerV.setLayoutManager(new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false));
        mRecyclerV.setNestedScrollingEnabled(false);
        mAdapter = new RecommendationsAdapter(this);
        mRecyclerV.setAdapter(mAdapter);
    }

    private void initEditText() {
        searchSubject = BehaviorSubject.create();
        mEditText.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId != 0 || event.getAction() == KeyEvent.ACTION_DOWN) {
                    onSubmitQuery();
                    View view = getActivity().findViewById(android.R.id.content);
                    if (view != null) {
                        InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
                        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
                    }
                    return true;
                }
                return false;
            }
        });

        String[] suggestions = getResources().getStringArray(R.array.suggestions);
        mEditText.setText(suggestions[new Random().nextInt(9)]);
        onSubmitQuery();
    }

    @OnClick(R.id.btnSearch)
    protected void onSubmitQuery() {
        if (mEditText.getText().toString().trim().length() > 0) {
            mPresenter.onSearchSources(mEditText.getText().toString().trim());
        }
    }

    @OnFocusChange(R.id.edit_query)
    protected void onFocusChanged(View view, boolean isFocused) {
        if (view.getId() == R.id.edit_query) {
            if (isFocused) {
                searchSubject.debounce(1000, TimeUnit.MILLISECONDS)
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(new Consumer<String>() {
                            @Override
                            public void accept(String s) throws Exception {
                                if (s.trim().length() > 0) {
                                    mPresenter.onSearchSources(s);
                                    mBtnSearch.setVisibility(View.GONE);
                                    mProgress.setVisibility(View.VISIBLE);
                                }
                            }
                        });
            }
        }
    }

    @OnTextChanged(R.id.edit_query)
    protected void onEtCurrencyChanged(CharSequence charSequence, int i, int i1, int i2) {
        if (!charSequence.toString().isEmpty()) {
            searchSubject.onNext(charSequence.toString());
        } else {
            mAdapter.onClearItems();
        }
    }

    @Override
    public void onAddNewSource() {
    }

    @Override
    public void onLoadSources(List<RssSource> rssSourceList) {
        mAdapter.onAddItems(rssSourceList);
    }

    @Override
    public void onNoContent(boolean isEmpty) {
        if (isEmpty) {
            mRecyclerV.setVisibility(View.GONE);
            tvNoContent.setVisibility(View.VISIBLE);
        } else {
            mRecyclerV.setVisibility(View.VISIBLE);
            tvNoContent.setVisibility(View.GONE);
        }
    }

    @Override
    public void onSourceClick(RssSource rssSource, boolean toDelete) {
        if (toDelete) {
            mPresenter.onRemoveSource(rssSource);
        } else {
            mPresenter.onAddNewSource(rssSource);
        }
        mListener.onUpdateRSS();
    }

    @Override
    public void onShowLoading() {
        mBtnSearch.setVisibility(View.GONE);
        mProgress.setVisibility(View.VISIBLE);
    }

    @Override
    public void onHideLoading() {
        mBtnSearch.setVisibility(View.VISIBLE);
        mProgress.setVisibility(View.GONE);
    }

    private AppComponent getAppComponent() {
        return ((App) getContext().getApplicationContext()).getAppComponent();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void onUnknownError(String errorMessage) {
        Toast.makeText(mContext, R.string.unknown_error, Toast.LENGTH_LONG).show();
    }

    @Override
    public void onTimeout() {
        Toast.makeText(mContext, R.string.timeout_error, Toast.LENGTH_LONG).show();
    }

    @Override
    public void onNetworkError() {
        Toast.makeText(mContext, R.string.network_error, Toast.LENGTH_LONG).show();
    }

    public interface OnFragmentInteractionListener {
        void onUpdateRSS();
    }
}

File: app/src/main/java/workshop/akbolatss/tagsnews/screen/recommendations/RecommendationsFragment.kt
package workshop.akbolatss.tagsnews.screen.recommendations

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.TextView
import android.widget.Toast
import java.util.Random
import java.util.concurrent.TimeUnit

import javax.inject.Inject

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.subjects.BehaviorSubject
import kotlinx.android.synthetic.main.fragment_recommendations.*
import workshop.akbolatss.tagsnews.R
import workshop.akbolatss.tagsnews.application.App
import workshop.akbolatss.tagsnews.di.component.DaggerSourcesComponent
import workshop.akbolatss.tagsnews.di.module.SourcesModule
import workshop.akbolatss.tagsnews.model.dao.RssSource
import workshop.akbolatss.tagsnews.screen.sources.SourcesPresenter
import workshop.akbolatss.tagsnews.screen.sources.SourcesView

/**
 * Fragment to search new RSS sources. Creates in #BoardActivity
 * @see BoardActivity
 */
class RecommendationsFragment : Fragment(), SourcesView, RecommendationsAdapter.RecommendationsListener {

    @Inject
    lateinit var mContext: Context

    @Inject
    lateinit var mPresenter: SourcesPresenter

    private var mAdapter: RecommendationsAdapter? = null
    private var mListener: OnFragmentInteractionListener? = null
    private var searchSubject: BehaviorSubject<String>? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_recommendations, container, false)
    }

    private fun initDagger() {
        DaggerSourcesComponent.builder()
                .appComponent((context!!.applicationContext as App).appComponent)
                .sourcesModule(SourcesModule(this))
                .build()
                .inject(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initDagger()
        initRV()
        initEditText()
    }

    private fun initRV() {
        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = LinearLayoutManager(activity, LinearLayoutManager.VERTICAL, false)
        recyclerView.isNestedScrollingEnabled = false
        mAdapter = RecommendationsAdapter(this)
        recyclerView.adapter = mAdapter
    }

    /**
     * Init TextInput View
     */
    private fun initEditText() {
        searchSubject = BehaviorSubject.create()
        etQuery.setOnEditorActionListener(TextView.OnEditorActionListener { _, actionId, event ->
            if (actionId != 0 || event.action == KeyEvent.ACTION_DOWN) {
                onSubmitQuery()
                val view = activity!!.findViewById<View>(android.R.id.content)
                if (view != null) {
                    val imm = activity!!.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                    imm.hideSoftInputFromWindow(view.windowToken, 0)
                }
                return@OnEditorActionListener true
            }
            false
        })

        etQuery.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {
            }
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                if (!s.toString().isEmpty()) {
                    searchSubject!!.onNext(s.toString())
                } else {
                    mAdapter!!.onClearItems()
                }
            }
        })

        val suggestions = resources.getStringArray(R.array.suggestions)
        etQuery.setText(suggestions[Random().nextInt(9)])
        onSubmitQuery()

        //Query for new result after typing within 1 second
        etQuery.onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
            if (hasFocus) {
                searchSubject!!.debounce(1000, TimeUnit.MILLISECONDS)
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe { s ->
                            if (s.trim { it <= ' ' }.isNotEmpty()) {
                                mPresenter.searchSources(s)
                                btnSearch.visibility = View.GONE
                                progressBar.visibility = View.VISIBLE
                            }
                        }
            }
        }
    }

    override fun onAddNewSource() {
        //not used in this class
    }

    /**
     * Submit query make search API call
     */
    private fun onSubmitQuery() {
        if (etQuery.text.toString().trim { it <= ' ' }.isNotEmpty()) {
            mPresenter.searchSources(etQuery.text.toString().trim { it <= ' ' })
        }
    }

    /**
     * Load items to RecyclerView.Adapter
     */
    override fun onLoadSources(rssSourceList: List<RssSource>) {
        mAdapter!!.onAddItems(rssSourceList)
    }

    override fun onNoContent(isEmpty: Boolean) {
        if (isEmpty) {
            recyclerView.visibility = View.GONE
            tvNoContent.visibility = View.VISIBLE
        } else {
            recyclerView.visibility = View.VISIBLE
            tvNoContent.visibility = View.GONE
        }
    }

    /**
     * Adds/Removes RSS source from DB
     * @param rssSource to save or delete
     */
    override fun onSourceClick(rssSource: RssSource, toDelete: Boolean) {
        if (toDelete) {
            mPresenter.removeSourceByLink(rssSource)
        } else {
            mPresenter.addNewSource(rssSource)
        }
        mListener!!.onUpdateRSS()
    }

    /**
     * Check in Database if exists
     */
    override fun checkIfExists(rssSource: RssSource): Boolean {
        return mPresenter.checkInSources(rssSource)
    }

    /**
     * Show loading state
     */
    override fun onShowLoading() {
        btnSearch.visibility = View.GONE
        progressBar.visibility = View.VISIBLE
    }

    /**
     * Hide loading state
     */
    override fun onHideLoading() {
        btnSearch.visibility = View.VISIBLE
        progressBar.visibility = View.GONE
    }

    /**
     * Attach listener
     */
    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            mListener = context
        } else {
            throw RuntimeException(context!!.toString() + " must implement OnFragmentInteractionListener")
        }
    }

    /**
     * Detach listener
     */
    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    override fun onUnknownError(errorMessage: String) {
        Toast.makeText(mContext, R.string.unknown_error, Toast.LENGTH_LONG).show()
    }

    override fun onTimeout() {
        Toast.makeText(mContext, R.string.timeout_error, Toast.LENGTH_LONG).show()
    }

    override fun onNetworkError() {
        Toast.makeText(mContext, R.string.network_error, Toast.LENGTH_LONG).show()
    }

    interface OnFragmentInteractionListener {
        fun onUpdateRSS()
    }
}
